extmod/network_ninaw10.c

```diff
@@ -124,21 +163,23 @@ STATIC void network_ninaw10_poll_connect(mp_sched_node_t *node) {
             }
         } else {
             // Will not attempt to reconnect if there's another error code set.
+            self->poll_enable = false;
             return;
         }
     }
 
     // Reschedule the connection polling code.
-    mp_sched_schedule_node(&mp_wifi_connpoll_node, network_ninaw10_poll_connect);
+    soft_timer_reinsert(&mp_wifi_poll_timer, 100);
 }
 
-STATIC mp_obj_t network_ninaw10_timer_callback(mp_obj_t none_in) {
-    if (MP_STATE_PORT(mp_wifi_sockpoll_list) != MP_OBJ_NULL && MP_STATE_PORT(mp_wifi_sockpoll_list)->len) {
-        mp_sched_schedule_node(&mp_wifi_sockpoll_node, network_ninaw10_poll_sockets);
+STATIC void network_ninaw10_timer_callback(soft_timer_entry_t *self) {
+    debug_printf("timer_callback() poll enable: %d\n", network_nina_wl_sta.poll_enable);
+    if (network_nina_wl_sta.poll_enable) {
+        mp_sched_schedule_node(&mp_wifi_poll_node, network_ninaw10_poll_connect);
+    } else if (!network_ninaw10_poll_list_is_empty()) {
+        mp_sched_schedule_node(&mp_wifi_poll_node, network_ninaw10_poll_sockets);
```

Comment:
You could have just one schedule node here and unconditionally schedule it, then in that callback put this logic:
```c
    if (network_nina_wl_sta.poll_enable) {
        network_ninaw10_poll_connect();
    }
    if (!network_ninaw10_poll_list_is_empty()) {
        network_ninaw10_poll_sockets();
    }
```
(The only reason this is needed is because soft-timer callbacks are executed at PendSV level, whereas you want to execute these poll_connect/poll_sockets functions at thread level.)

Reply:
I am using one schedule node. I'm not following, why schedule a callback if there's no work to be done ?

Reply:
Oh, I see, you are using just one node.  I missed that.

Is it always the case that, when `poll_enable` is true, only `network_ninaw10_poll_connect` needs to be called and never `network_ninaw10_poll_sockets`?

Maybe `poll_enable` could be renamed to `poll_connect_enable`?

Reply:
I can think of one corner case where the interface is connecting while the sockets poll list is Not empty. If you have sockets in the poll list, and get disconnected or call `disconnect()`, followed by `connect()` before the next poll is scheduled, then it's possible to get into that state.  However I don't think that's an issue, because on the next poll, after connecting, those sockets will return errors and get closed and removed from the list. But maybe it's better if the order is changed, to remove the sockets first if any, then poll connect:

```C
    if (!network_ninaw10_poll_list_is_empty()) {
        mp_sched_schedule_node(&mp_wifi_poll_node, network_ninaw10_poll_sockets);
    } else if (network_nina_wl_sta.poll_enable) {
        mp_sched_schedule_node(&mp_wifi_poll_node, network_ninaw10_poll_connect);
    }
```

> Maybe `poll_enable` could be renamed to `poll_connect_enable`?

This flag exists in both STA and AP interfaces, when set it means the interface should be polled for something, it can be extended later to poll the AP interface for something:

```C
    } else if (network_nina_wl_ap.poll_enable) {
        mp_sched_schedule_node(&mp_wifi_poll_node, network_ninaw10_poll_?);
    }
```

Reply:
I did some testing and it seems I made the wrong assumption about listening sockets being closed on disconnect(), they don't (I tested with CYW43 too). I also see what you meant about calling poll functions from the node's callback, but I think that can be avoided by just making sure that connect poll function will reinsert the timer at least one time after connecting, so the timer can call other poll functions (which in turn re-insert the timer). So basically STA/AP poll will take precedence over polling listening sockets.

I tested with with following simple script:
```Python
def accept_callback(listen_sock):
    pass

wifi_connect()

s = socket.socket()
s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
s.bind(socket.getaddrinfo("0.0.0.0", 8080)[0][4])
s.listen(1)
s.setsockopt(socket.SOL_SOCKET, 20, accept_callback)

time.sleep_ms(1000)
print("reconnecting")
wifi_connect()

while True:
    time.sleep_ms(1000)
```

The output shows that it initially polls the connection status (STA IF), then polls the listening socket, then the STA IF when reconnecting, and then back to listen socket:

```
timer_callback() poll status STA: 1 AP: 0 SOCKETS: 0
('192.168.50.53', '255.255.255.0', '192.168.50.1', '192.168.50.1')
socket_socket(2 1 0)
socket_setsockopt(54, 4)
socket_bind(54, 8080)
socket_listen(54, 1)
socket_setsockopt(54, 20)
...
timer_callback() poll status STA: 0 AP: 0 SOCKETS: 1
poll_sockets(54) -> flags 0
...
reconnecting
timer_callback() poll status STA: 1 AP: 0 SOCKETS: 1
timer_callback() poll status STA: 1 AP: 0 SOCKETS: 1
('192.168.50.53', '255.255.255.0', '192.168.50.1', '192.168.50.1')
timer_callback() poll status STA: 0 AP: 0 SOCKETS: 1
poll_sockets(54) -> flags 0
timer_callback() poll status STA: 0 AP: 0 SOCKETS: 1
...

Reply:
OK, that all sounds reasonable.  And good that you tested this reconnect scenario.